src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs

```diff
@@ -1272,10 +1272,20 @@ public static string GetModulePath(string currentProcessModulePath, string hklmM
 
         private static string UpdatePath(string path, string pathToAdd, ref int insertIndex)
         {
-            if (!string.IsNullOrEmpty(pathToAdd))
+            if (string.IsNullOrEmpty(pathToAdd))
             {
-                path = AddToPath(path, pathToAdd, insertIndex);
-                insertIndex = path.IndexOf(Path.PathSeparator, PathContainsSubstring(path, pathToAdd));
+                return path;
+            }
+
+            foreach (string entry in pathToAdd.Split(Path.PathSeparator))
```

Comment:
Hmm... such split is already in AddToPath()

Reply:
The problem is that `PathContainsSubstring ` returns `-1` causing an argument error with `IndexOf`.

https://github.com/PowerShell/PowerShell/blob/ba238808fbaed54db2a8ee1dd0e7c0013dae1875/src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs#L1278

`PathContainsSubstring` returns `-1` because it splits the path (yet again) by `Path.PathSeparator` and searches each entry for the current entry being added. This is fine if `UpdatePath` is adding a single entry but when that string itself contains the path separator it'll never find a match in `PathContainsSubstring`. The easy workaround I've added here is to just separate the entries in `UpdatePath` so that each check does work.

This probably needs a big rethink, the fact that there is different branches being run when PowerShell starts with or without the `PSModulePath` env var makes it super confusing. I'm not sure why we don't just have a `List<string>` of the paths and add them in the priority order already defined. We can easily check if something is already present and can avoid all the index searching and double up on checks that are currently used throughout the code.

Reply:
> I'm not sure why we don't just have a List<string>

I asked the team once, but it remained a mystery :-) Maybe it's just the principle of minimal changes. Although in this case there have been a lot of patches here already. The positive result of this is that some testing has been added, so you can take a chance and clean up this code.

Reply:
I would love to try and simplify the code but I'm unsure what the desired behaviour is when it comes to cleaning up things like duplicates and empty entries. Waiting on an answer from the WG also takes way too much time and if I'm being honest I doubt this one will actually get merged anytime soon which is a pity.

Maybe I'll just bite the bullet and do it because I really dislike the current behaviour and how it works.

Reply:
@iSazonov: PowerShell _ignores_ empty entries, both in `$env:PSModulePath` and in `$env:PATH`, on all platforms (by contrast, POSIX-compatible shells _do_ interpret an empty entry as including whatever the current directory is in the search path, and `cmd.exe` does so by default, irrespective of the `%PATH%` value).